README.md
cyclesafe\_app Cookbook
======================
Used to launch the cyclesafe app

Requirements
------------
### Programs
- Vagrant
- Chef
- Berkshelf
- ChefDK

### Cookbooks
- poise/application v4.1.6
- poise/application\_python v1.2.2
- poise/application\_nginx v2.0.0
- opscode-cookbooks/mysql v5.5.0
- opscode/poise
- opscode/sqlite

Attributes
----------
(will update)

Usage
-----
#### cyclesafe\_app::default
cyclesafe\_app::application
cyclesafe\_app::database

Contributing
------------
TODO: (optional) If this is a public cookbook, detail the process for contributing. If this is a private cookbook, remove this section.

e.g.
1. Fork the repository on Github
2. Create a named feature branch (like `add\_component\_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors: TODO: List authors


CHANGELOG.md
cyclesafe_chef CHANGELOG
=======================

0.1.0
-----
- [zpallin] - Initial release of cyclesafe_chef



recipes/database.rb



database_password = data_bag_item('passwords','database')['mysql']
db_name = node[:cyclesafe_chef][:db_name]

mysql_service db_name do
  version '5.5'
  port '3306'
  bind_address '127.0.0.1'
  initial_root_password database_password
  action [:create,:start]
end

user 'cyclesafe' do
  home node[:cyclesafe_chef][:directory]
  shell '/bin/bash'
  system true
end

directory node[:cyclesafe_chef][:directory] do
  action :create
  recursive true
  owner node[:cyclesafe_chef][:user]
  group node[:cyclesafe_chef][:group]
  mode 0755
end

template "#{node[:cyclesafe_chef][:directory]}/django_db.sql" do
  source 'django_db.sql.erb'
  variables ({
    :db_name => 'cyclesafe',
    :db_pass => database_password,
    :db_user => 'cyclesafe'
    })
  action :create
  owner node[:cyclesafe_chef][:user]
  group node[:cyclesafe_chef][:group]
  mode 0755
end

bash 'install_cyclesafe_user' do
  code "mysql -u root -p#{database_password} -h 127.0.0.1 < django_db.sql"
  cwd "#{node[:cyclesafe_chef][:directory]}"
  sensitive true
  user 'root'
end

# symbolic link to websocket, place where django expects it 
#     (because django isn't smart about mysql sockets yet)
symlink_dir = "/var/run/mysqld"
symlink_destination = "#{symlink_dir}/mysqld.sock"

# make sure mysqld directory exists
directory symlink_dir do
  action :create
  owner 'mysql'
  group 'mysql'
  mode 0777
  recursive true
  not_if {::File.exists?(symlink_dir)}
end

# then symlink the socket
bash 'symlink_mysql socket create' do
  code "ln -s /var/run/mysql-#{db_name}/mysqld.sock #{symlink_destination}"
  user 'root'
  sensitive true
  not_if {::File.exists?(symlink_destination)}
end

=begin
mysql_config 'default' do
  cookbook 'mysql'
  notifies :restart, 'mysql_service[default]'
  action :create
  source 'my.cnf.erb'
end

=begin
commands_list = [
  "sudo debconf-set-selections <<< 'mysql-server mysql-server/root_password password #{database_password}'",
  "sudo debconf-set-selections <<< 'mysql-server mysql-server/root_password_again password #{database_password}'",
  "apt-get -y install mysql-server"
  ]

bash 'install_mysql' do
  code "#{commands_list.join(';')}"
  user 'root'
  group 'root'
  not_if '-f /etc/init.d/mysql*'
end

=begin
mysql_cyclesafe_arr = [
  'create database if not exists cyclesafe;',
  "grant all on cyclesafe.* to 'cyclesafe'@'%' identified by '#{database_password}';",
  "flush privileges;"
  ]
mysql_cmd = "mysql -u root -p#{database_password} -e \"#{mysql_cyclesafe_arr.join(' ')}\""

bash 'install_cyclesafe_user' do
  code mysql_cmd
  sensitive true
  user 'root'
end

mysql_database 'cyclesafe' do
  connection( 
    :host => 'localhost',
    :username => 'cyclesafe',
    :password => database_password
  )
  action :create
end
=end


recipes/application.rb

# include defaults
include_recipe 'cyclesafe_chef'

# install dependencies
package 'libffi-dev'

# assign secret_key values
secret_key = data_bag_item('keys','secret_key')['key']
database_password = data_bag_item('passwords','database')['mysql']

# second secret_key escaped for supervisord pickiness!
secret_key_supervisord = "#{secret_key.gsub(/%/,'%%')}";

# easy variable references
sock_dir = "#{node[:cyclesafe_chef][:directory]}/run"
app_name = 'cyclesafe'
django_dir = "#{node[:cyclesafe_chef][:directory]}/current"
sock_file = "#{sock_dir}/#{app_name}.sock"
shared_dir = "#{node[:cyclesafe_chef][:directory]}/shared/env/"
wsgi_module = "#{app_name}.wsgi:application"
settings_module = "#{app_name}.settings"
log_level = node[:cyclesafe_chef][:log_level] || 'debug'
num_workers = 3

# create shared directory
directory "#{node[:cyclesafe_chef][:directory]}/shared" do
  user node[:cyclesafe_chef][:user]
  group node[:cyclesafe_chef][:group]
  mode '0775'
  recursive true
end

# set environment variable
ENV['SECRET_KEY'] = secret_key

# sock directory creation
directory sock_dir do
  action :create
  user node[:cyclesafe_chef][:user]
  group node[:cyclesafe_chef][:group]
  mode 0755
  recursive true
end

# install django
application app_name do
  path node[:cyclesafe_chef][:directory]
  owner node[:cyclesafe_chef][:user]
  group node[:cyclesafe_chef][:group]
  repository node[:cyclesafe_chef][:repository]
  revision node[:cyclesafe_chef][:revision]
  migrate true
  rollback_on_error false
  action :deploy

  django do
    requirements 'prod_requirements.txt'
    debug true
    packages ['gunicorn']
    settings_template 'settings.py.erb'
    
    database do
      database 'cyclesafe'
      adapter 'mysql'
      username 'cyclesafe'
      password database_password
      host '127.0.0.1'
      port 3306
    end
  end

  gunicorn do
    host node[:cyclesafe_chef][:hostname]
    app_module wsgi_module
    socket_path sock_file
    autostart true
    virtualenv shared_dir
    environment ({"SECRET_KEY"=>secret_key_supervisord})
  end

  nginx_load_balancer do
    server_name node[:cyclesafe_chef][:hostname]
    port 80
    application_socket ["#{sock_file} fail_timeout=0"]
    static_files '/static' => 'app/static'
  end
end

# supervisor reload resource
bash 'supervisor_reload' do
  code "supervisorctl restart #{app_name}"
  user 'root'
  action :nothing
end

# symlink current to local repo
symlink_directory = "#{node[:cyclesafe_chef][:directory]}/current"
symlink_destination = "#{node[:cyclesafe_chef][:directory]}/releases/local"
bash 'deploy_local' do
  user 'root'
  code "ln -sfnd #{symlink_destination} #{symlink_directory}"
  only_if { node[:instance_role] == 'vagrant' }
  notifies :run, 'bash[supervisor_reload]', :delayed
end
